feat: invoiceItem autofill

inpaceee opened the pull request:
closes: #1014 

fixes (besides ticket's goals):
- onBlur mutation call on production on invoiceDetails view;
- invoice create and update date handling;
- improves invoice table formatting on organization view;

inpaceee committed:
feat: invoiceItem autofill

inpaceee committed:
feat: cleanup

inpaceee committed:
feat: input fields, invoice item update mutation

inpaceee committed:
feat: new invoice handling, digits after decimal point, currency sign

inpaceee committed:
feat: autofill workHours from task if duration is deleted

inpaceee committed:
feat: more clean up, onBlur fix, comma instead of dot, defaultValue fix

inpaceee committed:
fix: typescript error upon adding new invoice item

inpaceee committed:
feat: clean up, new invoice item duration fix

inpaceee committed:
feat: adds an update invoice button

inpaceee committed:
feat: formatting of default value in the footer, clean up

inpaceee committed:
fix: hourly rate input field span element z index

inpaceee committed:
fix: organization handlers

inpaceee committed:
fix: comma for numbers, org handlers

inpaceee committed:
feat: number conversion clean up, dynamic footer, style for input field of type number

inpaceee committed:
fix: test

inpaceee committed:
fix: formatting of input fields, formatting of footer onBlur

inpaceee committed:
fix: test

YohannesTesfay commented:
- [x] The footer and the body input fields break if the Euro sign is present.
- [x] Error handling in the body if the input value is a string
- [x] Formatting in the body breaks upon entering a value
- [x] OnClick should match the onBlur behaviour in the footer

- [x] Date error handling in invoice create/ update
- [x] Invoice view in organization formatting

inpaceee committed:
feat: string input handling, formatting on input, onKeyDown handling

inpaceee committed:
feat: invoice create and update date error handling, organization view invoice table formatting, clean up

YohannesTesfay commented on apps/web/frontend/styles/globals.css:
```diff
@@ -2,3 +2,10 @@
 @tailwind base;
 @tailwind components;
 @tailwind utilities;
+@layer base {
+  input[type='number']::-webkit-inner-spin-button,
+  input[type='number']::-webkit-outer-spin-button {
+    -webkit-appearance: none;
+    margin: 0;
+  }
+}
```
Could you let me know what this does?

inpaceee replied:
Yes, it was Michael's suggestion to remove the arrows (used for iterating through values) from `InputField`s with `type="number"`. However, since then, I have removed the type attribute from the `InputField`s, which this fix was originally intended for. Let me know if you think it would be better to revert the change, but I believe the consensus was that the arrows were undesirable across the project.

YohannesTesfay replied:
Got it. Let's keep it then.

YohannesTesfay commented on apps/web/pages/organizations/[organizationId]/components/invoiceTable.tsx:
```diff
@@ -40,13 +40,25 @@ export const InvoiceTable = ({ invoices, organizationId }: InvoiceTableProps): J
+                    minimumFractionDigits: 2,
+                    maximumFractionDigits: 2,
+                    style: 'currency',
+                    currency: 'EUR',
```
I couldn't see the autocompletion for currency.

inpaceee replied:
Maybe that's a problem with your environment? https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/toLocaleString
![image](https://github.com/user-attachments/assets/5393b4dd-3251-4b37-bf71-1ef8970a9ef8)

YohannesTesfay replied:
I see.

YohannesTesfay commented on apps/web/pages/organizations/[organizationId]/components/invoiceTable.tsx:
```diff
@@ -40,13 +40,25 @@ export const InvoiceTable = ({ invoices, organizationId }: InvoiceTableProps): J
+                    maximumFractionDigits: 2,
+                  },
                 )}
               </td>
+              <td>
```
Could we also add the `Duration` column to give a full picture of the `Total amount`?
```suggestion
              </td>
              <td>
                {invoice.invoiceItems
                  .reduce((sum, invoiceItem) => sum + invoiceItem.duration, 0)
                  .toLocaleString(navigator.languages, {
                    minimumFractionDigits: 2,
                    maximumFractionDigits: 2,
                  })}
              </td>
              <td>
```

inpaceee replied:
I believe this *is* the duration column, just named differently. Unless you meant something else?

inpaceee replied:
What this table missing is the `Hourly rate`, which is tricky to implement, because different `Invoice item`s have different rate. So, we decided against it.

YohannesTesfay replied:
Got it.

YohannesTesfay commented on packages/backend/src/graphql/invoiceItem/mutations/invoiceItemUpdateMutationField.ts:
```diff
@@ -0,0 +1,26 @@
+        data: {
+          duration,
+          hourlyRate,
+          taskId: taskId.toString(),
```
Could we already make this a string in the `InvoiceItemInput`, like how the orgId in `projectInput` is?

inpaceee replied:
Hmm, do you think we should? If anything, I think we should change `organizationId` in `ProjectInput` to be `id()` as well, so that it's more semantically correct

YohannesTesfay replied:
I agree.

YohannesTesfay commented on apps/web/pages/organizations/[organizationId]/invoices/[invoiceId]/components/invoiceItemList.tsx:
```diff
@@ -61,21 +86,81 @@ export interface InvoiceItemListProps {
+  }
+  return newValue
+}
+
```
Could we extract these to some common component?

YohannesTesfay commented on apps/web/pages/organizations/[organizationId]/invoices/[invoiceId]/components/invoiceItemList.tsx:
```diff
@@ -61,21 +86,81 @@ export interface InvoiceItemListProps {
+  const [workHoursResult] = useQuery({
+    query: TaskWorkHoursQuery,
+    variables: {
+      id: invoiceItemsData[0]?.task.id || '',
```
Doesn't this `invoiceItemsData[0]` only check the first index?
What about then the rest of the `invoiceItemsData`?

YohannesTesfay commented:
Some points:
We first need to break the `invoiceItemList.tsx`  in to components. It’s very huge.

- [x] If all the tasks are added to the invoice, should we disable the inputs?
- Does it refresh the duration as soon as a task is selected from the list?
- Would it also be a good idea to filter out tasks that are already invoiced?

- [x] I think invoice start and end dates need validations. For example, it shouldn’t be possible to select/ create a past `to` date.
- Should we also consider fetching the projects/ tasks only from the range of the invoice dates?

inpaceee commented:
> Does it refresh the duration as soon as a task is selected from the list?

What do you mean by that? Fetching duration of a task into the duration input field in the footer, upon selecting it?

> Would it also be a good idea to filter out tasks that are already invoiced?

Right now I think it's out of the tickets scope. Right now you can only filter out by dates

> Should we also consider fetching the projects/ tasks only from the range of the invoice dates?

Currently, in the footer, you can select any task/project within the organization. However, when updating dates or clicking Update button, it only retrieves work hours that are within the invoice's start and end dates. Since work hours for any invoice item can be manually overridden, I believe we can leave it as is?

@YohannesTesfay

YohannesTesfay commented:
> > Does it refresh the duration as soon as a task is selected from the list?
>
> What do you mean by that? Fetching duration of a task into the duration input field in the footer, upon selecting it?

 Yes, to fetch the duration upon selecting the task.

> > Would it also be a good idea to filter out tasks that are already invoiced?
> 
> Right now I think it's out of the tickets scope. Right now you can only filter out by dates

Alright.

> > Should we also consider fetching the projects/ tasks only from the range of the invoice dates?
> 
> Currently, in the footer, you can select any task/project within the organization. However, when updating dates or clicking Update button, it only retrieves work hours that are within the invoice's start and end dates. Since work hours for any invoice item can be manually overridden, I believe we can leave it as is?

Sure.

inpaceee committed:
feat: invoiceItemUpdateInput mutation, validation, invociceItemListRow component, invoiceFormatUtils

inpaceee committed:
fix: tests and e2e-tests

inpaceee committed:
fix: test

YohannesTesfay reviewed (commented):
I tried creating a review branch: https://github.com/progwise/timebook/pull/1041 and assigned you as a reviewer.

YohannesTesfay commented:
> closes: #1014
> 
> fixes:
> 
> * onBlur mutation call on production on invoiceDetails view;
> * invoice create (_but not update for some reason_) date handling;
> * improves invoice table formatting on organization view;
> 
> * no implementation of deleting an invoice item currently;
> * if task is selected afterwards, one of the fields needs to be focused in order to trigger `onBlur`;
> * Update button, as well as updating dates of the invoice, does not update existing invoice items with workHours of their respective tasks, only adds missing ones;
> * error handling of invoiceWorkFrom to be after invoiceWorkUntil;
> 
> * tests

Closes: #1014 

Fixes:
- [x] onBlur mutation call on production on invoiceDetails view;
- [x] invoice create (*but not update for some reason*) date handling;
- [x] improves invoice table formatting on organization view;

Review branch fixes:
- [x] Error handling of invoiceWorkFrom to be after invoiceWorkUntil - Done in the review branch #1041 
       - The problem was the validation schema was using a different input type.
- [x]  If task is selected afterwards, one of the fields needs to be focused in order to trigger `onBlur`;
       - It doesn't make sense to add duration/ hourly rate before selecting a task - So I disabled them until a task is selected. 
- [x] Implementation of deleting an invoice item 
       - Done in #1043 
- [x] Update button, as well as updating dates of the invoice, does not update existing invoice items with workHours of their respective tasks, only adds missing ones;
       - Could also be seen as an advantage - the manual edits don't get lost. But I have still added filtering based on the invoice from - until dates. We have to discuss that though.

- Tests - ToDo

inpaceee committed:
Merge branch 'main' into invoice-item-autofill

YohannesTesfay committed:
Invoice-item-autofill-Review (#1041)

YohannesTesfay committed:
Invoice-item-delete (#1043)

* Improve invoice item autofill and validation

Some points addressed in this review pull request:
 - AuthScopes uses isAdminByOrganization both in Invoice and invoice items.
 - Filter out tasks/project that does not belong to the organization when updating an invoice.
 - Other implementations as well

* Implement delete invoice item

* Test for invoice item delete mutation field

* fix: errors from merge

* fix: organization handlers

* fix: organization handlers

* feat: UI suggestion

* feat: UI suggestion

---------

Co-authored-by: inpaceee <116308003+inpaceee@users.noreply.github.com>

inpaceee commented on packages/backend/src/graphql/invoiceItem/invoiceItemInput.ts:
```diff
@@ -5,6 +5,7 @@ import { builder } from '../builder'
 export const InvoiceItemInput = builder.inputType('InvoiceItemInput', {
   validate: { schema: invoiceItemInputValidations },
   fields: (t) => ({
+    organizationId: t.id({ required: false }),
```
Should this not be required since it is always present? Similar to `invoiceId` and `taskId`

inpaceee commented on packages/backend/src/graphql/invoiceItem/invoiceItem.ts:
```diff
@@ -16,9 +16,23 @@ export const InvoiceItem = builder.prismaObject('InvoiceItem', {
+        }),
     }),
   }),
 })
```
Can you elaborate on what this does please?

inpaceee commented on packages/backend/src/graphql/invoice/invoice.ts:
```diff
@@ -49,15 +49,6 @@ export const Invoice = builder.prismaObject('Invoice', {
-          orderBy: { task: { title: 'asc' } },
-        })
-      },
-    }),
```
Was this redundant?

inpaceee commented on apps/web/pages/organizations/[organizationId]/invoices/new.page.tsx:
```diff
@@ -57,8 +55,8 @@ const NewInvoicePage = (): JSX.Element => {
     } catch {}
   }
 
-  const handleCancel = async () => {
-    await router.push(`/organizations/${organizationId}`)
```
We have it implemented with async/await everywhere else, what is the reason to do / not do it this way?

inpaceee commented on apps/web/pages/organizations/[organizationId]/invoices/new.page.tsx:
```diff
@@ -41,7 +40,6 @@ const NewInvoicePage = (): JSX.Element => {
   } = useForm<InvoiceInput>({
     defaultValues: {
       organizationId: organizationId?.toString() ?? '',
-      invoiceDate,
```
Would like to have a feature at some point, to change the date of an invoice, but I guess we can get rid off it right now.

inpaceee commented on apps/web/pages/organizations/[organizationId]/invoices/invoiceInputSchema.tsx:
```diff
@@ -8,23 +8,24 @@ import { InvoiceInput } from '../../../../frontend/generated/gql/graphql'
   .string()
-  .refine((value) => value !== '____-__-__', 'enter a date')
-  .refine((value) => !value || isValid(parseISO(value)), 'invalid date')
+  .min(10, 'Enter a date')
+  .refine((value) => value !== '____-__-__', 'Enter a date')
+  .refine((value) => !value || isValid(parseISO(value)), 'Invalid date')
```
Is `.min(10, 'Enter a date')` necessary here?

inpaceee commented on apps/web/pages/organizations/[organizationId]/invoices/[invoiceId]/index.page.test.tsx:
```diff
@@ -37,8 +37,6 @@ it('should render invoice details', async () => {
-  expect(screen.getByDisplayValue('20.00')).toBeInTheDocument()
-  expect(screen.getByText('100.00')).toBeInTheDocument()
+  await screen.findByText('No tasks available')
+  expect(screen.getAllByDisplayValue('0.00')).toHaveLength(2)
```
What happened here

inpaceee commented on apps/web/pages/organizations/[organizationId]/invoices/[invoiceId]/components/invoiceItemListRow.tsx:
```diff
@@ -33,126 +34,98 @@ const InvoiceItemUpdateMutationDocument = graphql(`
-  const handleSubmitDurationHelper = async (invoiceItemData: Pick<InvoiceItemUpdateInput, 'duration'>) => {
-    const duration = invoiceItemData.duration ?? invoiceItem.duration
+  const router = useRouter()
+  const handleSubmitHelper = async (field: 'duration' | 'hourlyRate', value: number, oldValue: number) => {
```
```suggestion
  const handleSubmitHelper = async (field: keyof Pick<InvoiceItemUpdateInput, 'duration' | 'hourlyRate'>, value: number, oldValue: number) => {
```

inpaceee commented on apps/web/pages/organizations/[organizationId]/invoices/[invoiceId]/components/invoiceItemListRow.tsx:
```diff
@@ -33,126 +34,98 @@ const InvoiceItemUpdateMutationDocument = graphql(`
-    if (result.error) setError('hourlyRate', { message: 'Network error' })
+  const handleInputEvent = async (
+    event: React.FocusEvent<HTMLInputElement> | React.KeyboardEvent<HTMLInputElement>,
+    field: 'duration' | 'hourlyRate',
```
```suggestion
    field: keyof Pick<InvoiceItemUpdateInput, 'duration' | 'hourlyRate'>,
```

inpaceee commented on apps/web/pages/organizations/[organizationId]/invoices/[invoiceId]/components/invoiceItemList.tsx:
```diff
@@ -22,6 +23,8 @@ const InvoiceItemListInvoiceFragment = graphql(`
         id
         title
+        startDate
+        endDate
```
```suggestion
```

inpaceee commented on apps/web/pages/organizations/[organizationId]/invoices/[invoiceId]/components/invoiceItemList.tsx:
```diff
@@ -141,122 +141,131 @@ export const InvoiceItemList = ({ invoice }: InvoiceItemListProps): JSX.Element
+                : undefined
+              const invoiceStartDate = new Date(invoiceData.invoiceWorkFrom)
+              const invoiceEndDate = new Date(invoiceData.invoiceWorkUntil)
+              return taskStartDate <= invoiceEndDate && (!taskEndDate || taskEndDate >= invoiceStartDate)
```
Let's revert the filter, as we discussed yesterday
```suggestion
              .sort((a, b) => {
              const projectCompare = a.task.project.title.localeCompare(b.task.project.title)
              return projectCompare === 0 ? a.task.title.localeCompare(b.task.title) : projectCompare
```

inpaceee commented on apps/web/pages/organizations/[organizationId]/invoices/[invoiceId]/components/invoiceItemList.tsx:
```diff
@@ -22,6 +23,8 @@ const InvoiceItemListInvoiceFragment = graphql(`
       projects {
```
Feels wrong to have both `projects`  and `project` ...

inpaceee commented on apps/web/pages/organizations/[organizationId]/invoices/[invoiceId]/components/invoiceItemList.tsx:
```diff
@@ -141,122 +141,131 @@ export const InvoiceItemList = ({ invoice }: InvoiceItemListProps): JSX.Element
+                    event.target.value = getFormattedValue(value)
+                    handleFormSubmission(oldValues)
+                  }
+                }}
```
In my opinion, we shouldn't care about old value in the footer input field, as it should almost always be 0

inpaceee commented on apps/web/pages/organizations/[organizationId]/invoices/[invoiceId]/components/invoiceItemList.tsx:
```diff
@@ -141,122 +141,131 @@ export const InvoiceItemList = ({ invoice }: InvoiceItemListProps): JSX.Element
+                    event.target.value = getFormattedValue(value)
+                    handleFormSubmission(oldValues)
+                  }
+                }}
```
And same here

inpaceee commented on apps/web/pages/organizations/[organizationId]/invoices/[invoiceId]/components/invoiceFormatUtils.ts:
```diff
@@ -5,10 +5,10 @@ export const getFormattedValue = (value: number): string => {
   const newValue = Number(value)
   if (Number.isNaN(newValue)) {
-    return oldValue ?? 0
+    return Number.NaN
```
does this have same functionality?

inpaceee commented on apps/web/pages/organizations/[organizationId]/invoices/[invoiceId]/components/invoiceDetails.tsx:
Something is broken now about all the input fields here... They update the value but do not disappear (until refresh) on `onBlur`

inpaceee commented on apps/web/pages/organizations/[organizationId]/invoices/[invoiceId]/components/invoiceDetails.tsx:
```diff
@@ -100,43 +118,35 @@ export const InvoiceDetails = ({ invoice: invoiceFragment }: InvoiceDetailsProps
-                  handleSubmitHelper(editableDateField, { [editableDateField]: getValues(editableDateField) })
-                }}
-              />
+            <div className="flex flex-col gap-1">
```
```suggestion
```

inpaceee commented on apps/web/pages/organizations/[organizationId]/invoices/[invoiceId]/components/invoiceDetails.tsx:
```diff
@@ -100,43 +118,35 @@ export const InvoiceDetails = ({ invoice: invoiceFragment }: InvoiceDetailsProps
+                    handleSubmit(() => handleSubmitForm({ [editableDateField]: getValues(editableDateField) }))()
+                  }}
+                />
+              </div>
```
```suggestion
```

inpaceee commented on apps/web/pages/organizations/[organizationId]/invoices/[invoiceId]/components/invoiceDetails.tsx:
```diff
@@ -100,43 +118,35 @@ export const InvoiceDetails = ({ invoice: invoiceFragment }: InvoiceDetailsProps
             </div>
           )}
         />
-        {errors[editableDateField] && (
-          <ErrorMessage
-            name={editableDateField}
-            errors={errors}
-            as={<span role="alert" className="label-text-alt whitespace-nowrap text-error" />}
-          />
-        )}
```
No error message?

inpaceee commented on apps/web/pages/organizations/[organizationId]/invoices/[invoiceId]/components/invoiceDetails.tsx:
```diff
@@ -190,16 +218,14 @@ export const InvoiceDetails = ({ invoice: invoiceFragment }: InvoiceDetailsProps
             </button>
           </div>
-          <div>
+          <div className="text-left">
```
```suggestion
          <div>
```

inpaceee committed:
feat: addressing review review comments

inpaceee committed:
Merge branch 'main' into invoice-item-autofill

inpaceee committed:
Merge branch 'main' into invoice-item-autofill

inpaceee committed:
Merge branch 'invoice-item-autofill' of https://github.com/progwise/timebook into invoice-item-autofill

inpaceee committed:
Merge branch 'invoice-item-autofill' of https://github.com/progwise/timebook into invoice-item-autofill

inpaceee committed:
Merge branch 'invoice-item-autofill' of https://github.com/progwise/timebook into invoice-item-autofill

inpaceee commented:
Also, the way it works right now, is it doesn't fetch `workHours` of a `task` related to `invoiceItem` upon removing hours from respective `invoiceItem` anymore. Was this intended?

inpaceee commented:
It also looks like `organizationId` is needed in many places, but VSCode doesn't complain about it missing, nor are there any errors in the browser (it returns 200 with an error message inside which I found funny), for example here

![image](https://github.com/user-attachments/assets/ee77f278-7652-4f48-9f4b-43def10560de)

Perhaps change `InvoiceUpdateInput` to have `organizationId` as required?

inpaceee commented:
> >  If task is selected afterwards, one of the fields needs to be focused in order to trigger onBlur;
> - It doesn't make sense to add duration/ hourly rate before selecting a task - So I disabled them until a task is selected.

I'm still able to enter the values

![image](https://github.com/user-attachments/assets/adda32a6-4c5f-4a89-bc35-c175ee5314a7)

inpaceee committed:
fix: various fixes, adds missing orgId arguments, total amount calculation fix

inpaceee committed:
feat: various promise handling improvements, react elements and graphql mutations clean up, better invoice item list footer onBlur handling, invoiceItem.ts mutation clean up

YohannesTesfay reviewed (approved):
Great job!

inpaceee committed:
feat: backend tests
